File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.rb
# Definition for Node.
# class Node
#   attr_accessor :val, :next, :random
#   def initialize(val = 0)
#     @val = val
#	    @next = nil
#	    @random = nil
#   end
# end

# @param {Node} node
# @return {Node}
def copyRandomList(head)
  return head if !head
  
  orig, curr = [], head
  while curr
    orig << curr
    curr = curr.next
  end
  
  randoms = [0]*orig.length
  (0...orig.length).each { |i| randoms[i] = orig[i].random ? orig.index(orig[i].random) : -1 }
  
  ans, ans_arr = Node.new(head.val), []
  curr = ans
  (1...orig.length).each do |i|
    ans_arr << curr
    curr.next = Node.new(orig[i].val)
    curr = curr.next
  end
  ans_arr << curr
  
  randoms.each_with_index { |r, i| ans_arr[i].random = ans_arr[r] if r >= 0 }
  ans
end

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *   Val int
 *   Next *Node
 *   Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
  if head == nil {
    return head
  }
  orig, curr := []*Node{}, head
  for curr != nil {
    orig = append(orig, curr)
    curr = (*curr).Next
  }
  randoms := make([]int, len(orig))
  for i := 0; i < len(orig); i++ {
    randoms[i] = slices.Index(orig, (*orig[i]).Random)
  }

  ans := &Node{
    Val: (*head).Val,
  }
  ansArr := []*Node{}
  curr = ans
  for i := 1; i < len(orig); i++ {
    ansArr = append(ansArr, curr)
    (*curr).Next = &Node{
      Val: (*orig[i]).Val,
    }
    curr = (*curr).Next
  }
  ansArr = append(ansArr, curr)

  for i := 0; i < len(randoms); i++ {
    if randoms[i] >= 0 {
      (*ansArr[i]).Random = ansArr[randoms[i]]
    }
  }
  return ans
}